From 1d217247da8f6248beb0221f63d4f162fafa0aeb Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 28 Apr 2022 13:27:58 +0300
Subject: [PATCH] [SW-82716] common: add crc8 support

crc8 is used by pec digest in OOB support. This patch
add crc8 apis which will be used by OOB operations
to check integrity of data.

Change-Id: Ic04112f044f556871b1690fbc58855c19f52700b
---

diff --git a/common/crc8.c b/common/crc8.c
new file mode 100644
index 0000000..33b5ab2
--- /dev/null
+++ b/common/crc8.c
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <arcboot.h>
+#include <crc8.h>
+
+/* This code was copied from linux/lib/crc8.c */
+
+/**
+ * crc8_populate_msb - fill crc table for given polynomial in reverse bit order.
+ *
+ * @table:	table to be filled.
+ * @polynomial:	polynomial for which table is to be filled.
+ */
+void crc8_populate_msb(uint8_t table[CRC8_TABLE_SIZE], uint8_t polynomial)
+{
+	int i, j;
+	const uint8_t msbit = 0x80;
+	uint8_t t = msbit;
+
+	table[0] = 0;
+
+	for (i = 1; i < CRC8_TABLE_SIZE; i *= 2) {
+		t = (t << 1) ^ (t & msbit ? polynomial : 0);
+		for (j = 0; j < i; j++)
+			table[i+j] = table[j] ^ t;
+	}
+}
+
+/**
+ * crc8 - calculate a crc8 over the given input data.
+ *
+ * @table: crc table used for calculation.
+ * @pdata: pointer to data buffer.
+ * @nbytes: number of bytes in data buffer.
+ * @crc: previous returned crc8 value.
+ */
+uint8_t crc8(const uint8_t table[CRC8_TABLE_SIZE], uint8_t *pdata, size_t nbytes, uint8_t crc)
+{
+	/* loop over the buffer data */
+	while (nbytes-- > 0)
+		crc = table[(crc ^ *pdata++) & 0xff];
+
+	return crc;
+}
diff --git a/common/crc8.h b/common/crc8.h
new file mode 100644
index 0000000..acd15ca
--- /dev/null
+++ b/common/crc8.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _CRC8_H_
+#define _CRC8_H_
+
+#define CRC8_TABLE_SIZE		256
+
+void crc8_populate_msb(uint8_t table[CRC8_TABLE_SIZE], uint8_t polynomial);
+uint8_t crc8(const uint8_t table[CRC8_TABLE_SIZE], uint8_t *pdata, size_t nbytes, uint8_t crc);
+
+#endif
